testcases/kernel/syscalls/madvise/madvise08.c

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright (c) 2016 Richard Palethorpe <richiejp@f-m.fm>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+ * Check that memory marked with MADV_DONTDUMP is not included in a core dump
+ * and check that the same memory then marked with MADV_DODUMP is included in
+ * a core dump.
+ *
+ * In order to reliably find the core dump this test temporarily changes the
+ * system wide core_pattern setting. Meaning all core dumps will be sent to the
+ * test's temporary dir untill the setting is restored during cleanup.
+ *
+ * Test flow: map memory,
+ *	      write generated character sequence to memory,
+ *	      start child process,
+ *	      mark memory with MADV_DONTDUMP in child,
+ *	      abort child,
+ *	      scan child's core dump for character sequence,
+ *	      if the sequence is not found it is a pass otherwise a fail,
+ *	      do the opposite for MADV_DODUMP,
+ */
+
+#include "tst_test.h"
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <sys/prctl.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#ifndef MADV_DONTDUMP
+#define MADV_DONTDUMP 16
+#endif
+#ifndef MADV_DODUMP
+#define MADV_DODUMP 17
+#endif
+
+#define CORE_PATTERN "/proc/sys/kernel/core_pattern"
+#define CORE_FILTER "/proc/self/coredump_filter"
+
+struct test_case {
+	int fd;
+	void *fmem;
+	size_t fmemsize;
+	int ycount;
+	char *cpattern;
+	int dfd;
+};
+
+static struct test_case tc = {
+	.fd = 0,
+	.fmem = NULL,
+	.dfd = 0,
+	.ycount = 1280,
+	.cpattern = NULL,
+};
+
+static void setup(void)
+{
+	const char *dumpf = "dump-%p";
+	char cwd[1024];
+	char tmpcpattern[1048];
+	char *fmemc;
+	int i;
+	unsigned int filter;
+	struct rlimit limit;
+
+	tc.fmemsize = tc.ycount + 2;
+
+	SAFE_GETRLIMIT(RLIMIT_CORE, &limit);
+	if (limit.rlim_cur == 0) {
+		tst_brk(TCONF, "Core dump size resource limit is zero.");
+	} else if (limit.rlim_cur < 100000){
+		tst_res(TCONF, "Core dump size limit may be too low.");
+	}
+
+	switch (prctl(PR_GET_DUMPABLE)) {
+	case 0:
+		tst_brk(TCONF, "Process is not dumpable.");
+	case 1:
+		break;
+	default:
+		tst_brk(TBROK | TERRNO, "prctl(PR_GET_DUMPABLE)");
+	}
+
```

Comment:
I would not overcomplicate the test while trying to cope with all possible problems. If somebody does that and things break we can always add the check.

But as it is now most of the testcases expect pretty much standard Linux environment, even changing very basic things as umask will break some of them. So as far as I can tell most of the security measures are turned off before LTP is executed, since otherwise some amount of the testcases will not work correctly.